### PR TITLE
Add PayChan to recipient's owner directory

### PR DESCRIPTION
**Pull request by seelabs**

### CODE REVIEW

**Comment by seelabs** [comment]
I see the CI `g++: internal compiler error: Killed (program cc1plus)`. Will fix. The CI clang failures is the know clang 5 issue.

File: src/ripple/app/tx/impl/PayChan.cpp [hunk 1]
```diff
@@ -252,6 +263,16 @@ PayChanCreate::doApply()
         (*slep)[sfOwnerNode] = *page;
     }
 
+    // Add PayChan to the recipient's owner directory
+    if (ctx_.view().rules().enabled(fixPayChanRecipientOwnerDir))
+    {
+        auto page = dirAdd(ctx_.view(), keylet::ownerDir(dst), slep->key(),
```
Comment (scottschurr) [inline src/ripple/app/tx/impl/PayChan.cpp]: `page` could be `const` (and similarly on line 259).

File: src/ripple/app/tx/impl/PayChan.cpp [hunk 2]
```diff
@@ -139,6 +139,17 @@ closeChannel (
         }
     }
 
+    // Remove PayChan from recipient's owner directory, if present.
+    if (auto const page = (*slep)[~sfDestinationNode];
+        page && view.rules().enabled(fixPayChanRecipientOwnerDir))
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/ripple/app/tx/impl/PayChan.cpp]: I agree that we should never see a paychan in the recipient's owner dir without the amendment, so it's redundant. Still, I'd still like to keep the amendment (tho I will remove it if others feel strongly).

File: src/test/app/PayChan_test.cpp [hunk 3]
```diff
@@ -1136,6 +1152,101 @@ struct PayChan_test : public beast::unit_test::suite
...
+                env.current()->rules().enabled(fixPayChanRecipientOwnerDir));
+            // close the channel after the amendment activates
+            env(claim(bob, chan), txflags(tfClose));
+            BEAST_EXPECT(!channelExists(*env.current(), chan));
+            BEAST_EXPECT(!inOwnerDir(*env.current(), alice, chanSle));
+            BEAST_EXPECT(ownerDirCount(*env.current(), alice) == 0);
+            BEAST_EXPECT(!inOwnerDir(*env.current(), bob, chanSle));
+            BEAST_EXPECT(ownerDirCount(*env.current(), bob) == 0);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/test/app/PayChan_test.cpp]: Thanks!

File: src/ripple/app/tx/impl/PayChan.cpp [hunk 4]
```diff
@@ -139,6 +139,17 @@ closeChannel (
...
+    {
+        auto const dst = (*slep)[sfDestination];
+        if (!view.dirRemove(keylet::ownerDir(dst), *page, key, true))
+        {
+            return tefBAD_LEDGER;
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/ripple/app/tx/impl/PayChan.cpp]: Agreed. Fixed in 225b9ca2d `[fold] Write to log on bad ledger`

File: src/test/app/PayChan_test.cpp [hunk 5]
```diff
@@ -1136,6 +1152,101 @@ struct PayChan_test : public beast::unit_test::suite
...
+            ripple::Dir ownerDir(view, keylet::ownerDir(acc.id()));
+            return std::find(ownerDir.begin(), ownerDir.end(), chan) != ownerDir.end();
+        };
+
+        auto ownerDirCount = [](ReadView const& view,
+                                Account const& acc) -> std::size_t
+        {
+            ripple::Dir ownerDir(view, keylet::ownerDir(acc.id()));
+            return std::distance(ownerDir.begin(), ownerDir.end());
+
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/test/app/PayChan_test.cpp]: Fixed in 613db91af `[fold] Add const in some places & fix typo`

File: src/test/app/PayChan_test.cpp [hunk 6]
```diff
@@ -1136,6 +1152,101 @@ struct PayChan_test : public beast::unit_test::suite
...
+            BEAST_EXPECT(ownerDirCount(*env.current(), bob) == 0);
+        }
+
+        {
+            // Test removing paychans created before adding to the recipient's
+            // owner directory
+            Env env(
+                *this, supported_amendments() - fixPayChanRecipientOwnerDir);
+            env.fund (XRP (10000), alice, bob);
+            // create the channels before the amendment activates
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/test/app/PayChan_test.cpp]: Fixed in 613db91af `Add const in some places & fix typo`

File: src/test/app/PayChan_test.cpp [hunk 7]
```diff
@@ -1136,6 +1152,101 @@ struct PayChan_test : public beast::unit_test::suite
...
+
+        auto const alice = Account("alice");
+        auto const bob = Account("bob");
+        auto const settleDelay = 100s;
+        auto const pk = alice.pk();
+
+        auto inOwnerDir = [](ReadView const& view,
+                             Account const& acc,
+                             std::shared_ptr<SLE const> const& chan) -> bool {
+            ripple::Dir ownerDir(view, keylet::ownerDir(acc.id()));
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/test/app/PayChan_test.cpp]: Fixed in 613db91af `Add const in some places & fix typo`

File: src/test/app/PayChan_test.cpp [hunk 8]
```diff
@@ -1136,6 +1152,101 @@ struct PayChan_test : public beast::unit_test::suite
...
+            return std::find(ownerDir.begin(), ownerDir.end(), chan) != ownerDir.end();
+        };
+
+        auto ownerDirCount = [](ReadView const& view,
+                                Account const& acc) -> std::size_t
+        {
+            ripple::Dir ownerDir(view, keylet::ownerDir(acc.id()));
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/test/app/PayChan_test.cpp]: Fixed in 613db91af `Add const in some places & fix typo`

File: src/ripple/app/tx/impl/PayChan.cpp (hunk 1 above)
Comment (seelabs) [reply src/ripple/app/tx/impl/PayChan.cpp]: Fixed in 613db91af `Add const in some places & fix typo`

**Comment by seelabs** [comment]
The internal compile error appears to be a jenkins only issue. I can't repro with a local gcc 7.3. Apparently jenkins was having issues yesterday. I don't believe the error is related to this PR (but of course will keep an eye on it).

**Comment by scottschurr** [comment]
Still :+1: from me.

File: src/ripple/app/tx/impl/PayChan.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (miguelportilla) [reply src/ripple/app/tx/impl/PayChan.cpp]: If you do keep it, you may want to consider evaluating the amendment before constructing `page`.
e.g.
```
if (view.rules().enabled(fixPayChanRecipientOwnerDir))
{
    if (auto const page = (*slep)[~sfDestinationNode])
    {
        ...
    }
}
```

File: src/test/app/PayChan_test.cpp [hunk 9]
```diff
@@ -1136,6 +1152,101 @@ struct PayChan_test : public beast::unit_test::suite
...
+            env(claim(bob, chan), txflags(tfClose));
+            BEAST_EXPECT(!channelExists(*env.current(), chan));
+            BEAST_EXPECT(!inOwnerDir(*env.current(), alice, chanSle));
+            BEAST_EXPECT(ownerDirCount(*env.current(), alice) == 0);
+            BEAST_EXPECT(!inOwnerDir(*env.current(), bob, chanSle));
+            BEAST_EXPECT(ownerDirCount(*env.current(), bob) == 0);
+        }
+
+        {
+            // Test with adding the paychan to the recipients owner directory
```
Comment (miguelportilla) [inline src/test/app/PayChan_test.cpp]: nit: recipient's

File: src/test/app/PayChan_test.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (miguelportilla) [reply src/test/app/PayChan_test.cpp]: Sub-Sub-nit: blank line persists after[613db91](https://github.com/ripple/rippled/commit/613db91afa770668e35011c68720d223f5a22552)

File: src/test/app/PayChan_test.cpp [hunk 10]
```diff
@@ -1136,6 +1152,101 @@ struct PayChan_test : public beast::unit_test::suite
...
+            env(create(alice, bob, XRP(1000), settleDelay, pk));
+            auto const [chan, chanSle] = channelKeyAndSle(*env.current(), alice, bob);
+            BEAST_EXPECT(inOwnerDir(*env.current(), alice, chanSle));
+            BEAST_EXPECT(ownerDirCount(*env.current(), alice) == 1);
+            BEAST_EXPECT(!inOwnerDir(*env.current(), bob, chanSle));
+            BEAST_EXPECT(ownerDirCount(*env.current(), bob) == 0);
+            env.enableFeature(fixPayChanRecipientOwnerDir);
+            env.close();
+            BEAST_EXPECT(
+                env.current()->rules().enabled(fixPayChanRecipientOwnerDir));
```
Comment (miguelportilla) [inline src/test/app/PayChan_test.cpp]: After enabling, consider testing if bob now has in his owner directory and the count.
e.g.
```
            BEAST_EXPECT(inOwnerDir(*env.current(), bob, chanSle));
            BEAST_EXPECT(ownerDirCount(*env.current(), bob) == 1);
```

File: src/test/app/PayChan_test.cpp [hunk 11]
```diff
@@ -18,13 +18,15 @@
...
 #include <ripple/basics/chrono.h>
+#include <ripple/ledger/Directory.h>
 #include <ripple/protocol/Feature.h>
 #include <ripple/protocol/Indexes.h>
 #include <ripple/protocol/jss.h>
 #include <ripple/protocol/PayChan.h>
 #include <ripple/protocol/TxFlags.h>
 #include <test/jtx.h>
 
+
```
Comment (miguelportilla) [inline src/test/app/PayChan_test.cpp]: nit: Extra line?

File: src/ripple/app/tx/impl/PayChan.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (seelabs) [reply src/ripple/app/tx/impl/PayChan.cpp]: I hear you. But I think it's OK as-is. Here's my thinking:
1) There will be a small windows of time where the fix is not enabled. We can avoid creating a page in this case.
2) After the fix is enabled, we'll have to create the page from then until forever.

Having said that, I don't really object to checking the amendment before constructing the page and will do so if people feel strongly.

Note: I spoke with @miguelportilla in a DM and he's OK with this. I just wanted to leave a note for other reviewers.

File: src/test/app/PayChan_test.cpp (hunk 9 above)
Comment (seelabs) [reply src/test/app/PayChan_test.cpp]: Fixed in 499e9e0e3 `Add `close` after channel create and fix typos`

File: src/test/app/PayChan_test.cpp (hunk 11 above)
Comment (seelabs) [reply src/test/app/PayChan_test.cpp]: Fixed in c424ff1bd `TRIVIAL fixup whitespace`

File: src/test/app/PayChan_test.cpp (hunk 10 above)
Comment (seelabs) [reply src/test/app/PayChan_test.cpp]: So this is really interesting. It's intended that Bob's owner count should be zero and the channel is not part of the owner directory. However, as written, this isn't true. I need to `close` after creating the channel or the open ledger runs with one set of amendments and when closing runs with another set of amendments.

I fixed this in `499e9e0e3 Add `close` after channel create and fix typos`

**Pull request closed without merging**